content/backend/graphql-scala/7-relations.md

```diff
@@ -0,0 +1,457 @@
+
+Because of relation with `User`, `Link` has two fields that could be used to fetch it by: `id` and `postedBy`.
+The first is default and used in normal way, the second is used if we want to find link for proper user.
+
+<Instruction>
+
+Add relation to able to find `Link`s by userId.
+
+```scala
+
+val linkByUserRel = Relation[Link, Int]("byUser", l => Seq(l.postedBy))
+```
+
+</Instruction>
+
+This kind Relation is of type `SimpleRelation` and has only two arguments: the first is name, and the second is a function which somehow extracts sequence of user id's from link entity. In our case is super easy, because `postedBy` has such id. All we need to do is wrap it into the sequence.
+
+Now we have to add this relation to the fetcher, to do this, we have to use `Fetcher.rel` function instead previously used `apply`
+
+<Instruction>
+
+Change `linksFetcher` to the following:
+
+```scala
+val linksFetcher = Fetcher.rel(
+    (ctx: MyContext, ids: Seq[Int]) => ctx.dao.getLinks(ids),
+    (ctx: MyContext, ids: RelationIds[Link]) => ctx.dao.getLinksByUserIds(ids(linkByUserRel))
+  )
+
+```
+
+</Instruction>
+
+What do we have here? As I mentioned above, now we're using `.rel` function. It needs the second function to be passed as the argument. This function is for fetching related data from datasource. In our case it uses a function `getLinksByUserIds` what we have to add to our dao.
+
+
+<Instruction>
+
+In `DAO` class add a function:
+
+```scala
+def getLinksByUserIds(ids: Seq[Int]): Future[Seq[Link]] = {
+    db.run {
+      Links.filter(_.postedBy inSet ids).result
+    }
+}
+```
+
+</Instruction>
+
+Actually we've simplified the code above a little. When you will thing more about the part `ctx.dao.getLinksByUserIds(ids(linkByUserRel)` you can ask "And what if link has two relations? Does `getLinkByUserIds` could be replaced by another function?" Be patient, such case will be covered later in this chapter.
+In our case we have only one relation, so we can retrieve all `userId`'s by calling `ids(linkByUserRel)` functions.
+
+### Add fields to GraphQL Objects
+
+Let's begin from `LinkType`. `Link` already has `postedBy` field, but for now it's only an `Int` and we need entire user.
+
+<Instruction>
+
+Add `ReplaceField` type class to the `LinkType` constructor.
+
+```scala
+ReplaceField("postedBy",
+      Field("postedBy", UserType, resolve = c => usersFetcher.defer(c.value.postedBy))
+)
+```
+
+</Instruction>
+
+As you can see `resolve` uses basic users fetcher on this side.
+
+In the same way we will change the `UserType` but we won't replace any existing field, so we have to use `AddField` instead.
+
+<Instruction>
+
+```scala
+AddFields(
+      Field("links", ListType(LinkType), resolve = c =>  linksFetcher.deferRelSeq(linkByUserRel, c.value.id))
+)
+```
+
+</Instruction>
+
+Now you can see, there is called another fetcher's function. All `.deferRel...` functions needs two arguments instead of one. We have to add relation object as first.
```

Comment:
`as first` -> drop `as first` or `We need to pass the relation object as first argument`